src/main/java/com/cloudbees/hudson/plugins/folder/AbstractFolder.java

```diff
@@ -565,8 +566,17 @@ public String call(I item) {
                     float percentage = 100.0f * jobEncountered.incrementAndGet() / Math.max(1, jobTotal.get());
                     long now = System.currentTimeMillis();
                     if (loadingTick == 0) {
+                        // Buy ourselves two tick intervals to complete the first tick + some slop.
+                        Jenkins.get().getLifecycle().onExtendTimeout(2 * TICK_INTERVAL, TimeUnit.MILLISECONDS);
+
                         loadingTick = now;
                     } else if (now - loadingTick > TICK_INTERVAL) {
+                        // If we're still making progress loading jobs, buy ourselves another tick.
+                        if (jobEncountered.get() > jobEncounteredAtLastTick.get()) {
+                            Jenkins.get().getLifecycle().onExtendTimeout(TICK_INTERVAL, TimeUnit.MILLISECONDS);
+                        }
+                        jobEncounteredAtLastTick.set(jobEncountered.get());
```

Comment:
There is probably some idiomatic way to do the compare and set atomically but I suppose we do not care much here since it is only a heuristic.

(Are these counters actually used from concurrent threads? Predates 869ef0ce662d643aa16f3ce55150a34f0c7b41d2 and I no longer recall!)

Reply:
> There is probably some idiomatic way to do the compare and set atomically

If there is such an idiomatic way, I cannot think of it.

> Are these counters actually used from concurrent threads?

I am not familiar with this code, but reading it raised more questions than answers in my mind (i.e., it smelled wrong to me) so I decided not to think too hard about it.